manager.go

```diff
@@ -457,6 +390,117 @@ func (m *Manager) moduleChanged(
 	return changed, fmt.Sprintf("module %q changed because %s", mod.Source, why), nil
 }
 
+// listChangedFiles lists all changed files in the dir directory.
+func listChangedFiles(dir string, gitBaseRef string) ([]string, error) {
+	logger := log.With().
+		Str("action", "listChangedFiles()").
+		Str("path", dir).
+		Logger()
+
+	logger.Trace().
+		Msg("Get dir info.")
+	st, err := os.Stat(dir)
+	if err != nil {
+		return nil, fmt.Errorf("stat failed on %q: %w", dir, err)
+	}
+
+	logger.Trace().
+		Msg("Check if path is dir.")
+	if !st.IsDir() {
+		return nil, fmt.Errorf("is not a directory")
+	}
+
+	logger.Trace().
+		Msg("Create git wrapper with dir.")
+	g, err := git.WithConfig(git.Config{
+		WorkingDir: dir,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Trace().
+		Msg("Check if path is git repo.")
+	if !g.IsRepository() {
+		return nil, fmt.Errorf("the path \"%s\" is not a git repository", dir)
+	}
+
+	err = checkRepoIsClean(g)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Trace().
+		Msg("Get commit id of git base ref.")
+	baseRef, err := g.RevParse(gitBaseRef)
+	if err != nil {
+		return nil, fmt.Errorf("getting revision %q: %w", gitBaseRef, err)
+	}
+
+	logger.Trace().
+		Msg("Get commit id of HEAD.")
+	headRef, err := g.RevParse("HEAD")
+	if err != nil {
+		return nil, fmt.Errorf("getting HEAD revision: %w", err)
+	}
+
+	if baseRef == headRef {
+		return []string{}, nil
+	}
+
+	logger.Trace().
+		Msg("Find common commit ancestor of HEAd and base ref.")
+	mergeBaseRef, err := g.MergeBase("HEAD", baseRef)
+	if err != nil {
+		return nil, fmt.Errorf("getting merge-base HEAD main: %w", err)
+	}
+
+	if baseRef != mergeBaseRef {
+		return nil, fmt.Errorf("main branch is not reachable: main ref %q can't reach %q",
+			baseRef, mergeBaseRef)
+	}
+
+	return g.DiffNames(baseRef, headRef)
+}
+
+func checkRepoIsClean(g *git.Git) error {
+	logger := log.With().
+		Str("action", "checkRepoIsClean()").
+		Logger()
+
+	logger.Debug().
+		Msg("Get list of untracked files.")
+	untracked, err := g.ListUntracked()
+	if err != nil {
+		return fmt.Errorf("listing untracked files: %v", err)
+	}
+
+	logger.Debug().
+		Msg("Get list of uncommitted files in dir.")
+	uncommitted, err := g.ListUncommitted()
+	if err != nil {
+		return fmt.Errorf("listing uncommitted files: %v", err)
+	}
+
+	if len(untracked) > 0 {
+		logger.Warn().
+			Strs("files", untracked).
+			Msg("repository has untracked files")
```

Comment:
in case of run this should be error level how will this work?

Reply:
As it's the caller who decides if it's a warning or an error, we can change those WARN to TRACE and return the error and let the caller log a WARN or an ERR/FTL. 

Currently it shows this:
![Screenshot_2022-01-11_10-01-48](https://user-images.githubusercontent.com/251478/148921943-77f2cda7-4939-43b5-984d-6d70813b48e8.png)